src/openms/source/ANALYSIS/TARGETED/MetaboTargetedAssay.cpp

```diff
@@ -0,0 +1,498 @@
+      if (method_consensus_spectrum &&index.size()>= 2)
+      {
+        // transform to binned spectra
+        vector <BinnedSpectrum> binned;
+        vector <MSSpectrum> similar_spectra;
+        MSExperiment exp;
+        const BinnedSpectrum binned_highest_int(highest_precursor_int_spectrum,
+                                                BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES,
+                                                false,
+                                                1,
+                                                BinnedSpectrum::DEFAULT_BIN_OFFSET_HIRES);
+
+        // calculation of contrast angle (cosine simiarity)
+        for (auto index_it = index.begin(); index_it != index.end(); ++index_it)
+        {
+          const MSSpectrum &spectrum = spectra[*index_it];
+
+          const BinnedSpectrum binned_spectrum(spectrum,
+                                               BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES,
+                                               false,
+                                               1,
+                                               BinnedSpectrum::DEFAULT_BIN_OFFSET_HIRES);
+
+          BinnedSpectralContrastAngle bspa;
+          double cosine_sim = bspa(binned_highest_int, binned_spectrum);
+          if (cosine_sim > cosine_sim_threshold)
+          {
+            similar_spectra.push_back(spectrum);
+            exp.addSpectrum(spectrum);
+          }
+        }
+
+        // at least 2 spectra with high consine similarity necessary
+        // fallback to highest precursor intensity spectrum (see above)
+        if (similar_spectra.size()> 1)
+        {
+          // calculate consensus spectrum
+          exp.sortSpectra();
+
+          SpectraMerger merger;
+          Param p;
+          p.insert("",SpectraMerger().getDefaults());
+          p.setValue("precursor_method:mz_tolerance", precursor_mz_distance);
+          p.setValue("precursor_method:rt_tolerance", precursor_rt_tol * 2);
+          merger.setParameters(p);
+
+          // all MS spectra should have the same precursor
+          merger.mergeSpectraPrecursors(exp);
+
+          // check if all precursors have been merged if not use highest intensity precursor
+          if (exp.getSpectra().size()< 2)
+          {
+            transition_spectrum = exp.getSpectra()[0];
+          }
+        }
+      }
+
+      // transition calculations
+      // calculate max intensity peak and threshold
+      float max_int = 0.0;
+      float min_int = std::numeric_limits<float>::max();
+
+      // sort intensity in MS2 spectrum to extract transitions
+      transition_spectrum.sortByIntensity(true);
+
+      // filter out the precursors if they are in the ms2 spectrum;
+      if (exclude_ms2_precursor)
+      {
+        for (auto spec_it = transition_spectrum.begin(); spec_it != transition_spectrum.end(); ++spec_it)
+        {
+          if (transition_spectrum.getPrecursors()[0].getMZ() == spec_it->getMZ())
```

Comment:
why exact matching of floating point number?
```suggestion
          if ( abs(transition_spectrum.getPrecursors()[0].getMZ() - spec_it->getMZ()) < 1e-3)
```

Reply:
That makes sense...